### ISSUE TITLE
Reproducing throughput numbers for pretrained models

### ISSUE THREAD

**Issue by jwkirchenbauer** [UKPLab/sentence-transformers#2341]
Hi @nreimers!

TLDR: was the throughput number (14200 sent/sec) for the `all-MiniLM-L6-v2` on a single V100 gpu computed using very short test sentences?

---

I'm trying to use this library to encode large numbers of texts (eventually many many millions) and happen to be running some of this on V100 gpus and I noticed that "sentences/second" throughput numbers were reported for the models in the pretrained models table. Since I am using the same (older) GPUs for this effort, I thought that the performance of my code should be roughly equivalent to those numbers, but on first look it seemed like it was not.

Using some custom code that applies the sentence transformers encoding function to a huggingface dataset using the multiprocessing map feature, using a whole node of 8 x V100 gpus, for a set of "sentences" that are 50 white space separated english words long, I get about 15k embeddings/second. Since the table reports 14200 sentences/second on _**one** V100_ I was concerned...

<img width="1243" alt="Screenshot 2023-10-30 at 8 51 58 PM" src="https://github.com/UKPLab/sentence-transformers/assets/74567163/c2824293-da6d-4106-acb0-d3d72b6e3913">

I created a minimum example like so:
```
from sentence_transformers import SentenceTransformer

all_sentences = ["This framework generates embeddings for each input sentence"]*1_000_000
model = SentenceTransformer('all-MiniLM-L6-v2')

all_embeddings = model.encode(all_sentences, device="cuda", show_progress_bar=True, convert_to_numpy=False, batch_size=256)
```
This is seemingly a very generous setting because the "sentence" is in fact a sentence and is quite short (I actually grabbed the example sentence from the documentation). I got a runtime of 68 seconds on the one GPU for a throughput of ~14700 sentences/second.
But then I tried again with a set of slightly longer sequences:
```
all_sentences = ["This is a test sequence with a short sequence length of approximately one sentence. And here is another sentence with something similar to the length of the first. Finally, here is one more sentence to round out the collection of three sentences."]*1_000_000
```
and got a runtime of 122 seconds on the single card for a throughput of closer to ~8000 sentences/second.

Then, circling back, for a sample of my own data (sequences of 50 whitespace sep'd tokens):
```
print(len(all_sentences))
> 3469517
print(all_sentences[:3])
> ['The woman who died after falling from a bridge over the A21 has been identified as a Sevenoaks mum. Marta Kendle, 37, fell from the Gracious Lane bridge on the morning of February 19. Police were called to the carriageway around 6.10am and the road was promptly closed in both',
 'around 6.10am and the road was promptly closed in both directions. Despite paramedics best efforts, Marta, who was originally from Poland, was pronounced dead at the scene. Kent and Medway Coroners office have confirmed an inquest into her death will open on Wednesday (February 27). Tributes to the mum were',
 'open on Wednesday (February 27). Tributes to the mum were left at the scene and on social media. Friend, Jodi Cahill posted on Facebook: "I will certainly remember you. I am sorry we did not see how lost and alone you felt. "Be at peace dear Marta." A floral tribute']
```
I tried the same setup on the single card and it ran in  ~19  mins for a throughput of only 3000 sequences/second.

Finally, using the `start_multi_process_pool` + `encode_multi_process` pattern with a batch_size of 256 I got a throughput with _all 8_ cards of ~15000 sequences/second. This compares almost exactly to my totally different 8 gpu code which also runs the same sequences with about ~15000 sequences/second.

So... what was the approximate setup for producing the throughput numbers in the pretrained models table? 
**_Was it using a set of very short sentences as input?_** 
This would seem like the only way to achieve the 14200 sentences/second number cited in the table for this model on a single v100 gpu.

If not, (yikes for me) are there any pointers for maximizing the throughput of encoding large numbers of sequences?

**Comment by tomaarsen** [comment]
I'm not 100% sure on this, but it seems very plausible that [this script](https://github.com/UKPLab/sentence-transformers/blob/master/examples/evaluation/evaluation_inference_speed.py) was used to determine the throughput.

If I use NLTK's `word_tokenize` to tokenize these 100k sentences, then I get the following stats:
* min length: 2 words
* max length: 82 words
* median length: 12 words
* mean length: 13.32681 words

Here are some sample sentences with 12 words:
* 'A woman pushes a man in a wheelchair at the beach.'
* 'A man and a woman are sitting at a table drinking.'
* 'Two men, on smoking, talk outside a street-side restaurant.'
* 'A motorcycle is lying on the ground near a fire engine.'

So, yes, I think Nils used short sentences in his experiments.

- Tom Aarsen

**Comment by jwkirchenbauer** [comment]
Awesome! This seems to check out with what I guessed based on my tests. Thanks so much for looking into it and computing those statistics 🙂
